packages/aws-cdk-lib/core/test/aspect.prop.test.ts

```diff
@@ -0,0 +1,748 @@
+            throw new Error(
+              `Aspect ${a.aspect}@${aPrio} at ${a.index} should have been before ${b.aspect}@${bPrio} at ${b.index}, but was after`,
+            );
+          }
+        });
+      }, stabilizeAspects)(app);
+    }),
+  ),
+);
+
+test('visitLog is nonempty', () =>
+  fc.assert(
+    fc.property(appWithAspects(), fc.boolean(), (app, stabilizeAspects) => {
+      afterSynth((testApp) => {
+        expect(testApp.visitLog).not.toEqual([]);
+      }, stabilizeAspects)(app);
+    }),
+  ),
+);
+
+//////////////////////////////////////////////////////////////////////
+//  Test Helpers
+
+function afterSynth(block: (x: PrettyApp) => void, stabilizeAspects: boolean) {
+  return (app) => {
+    let asm;
+    try {
+      asm = app.cdkApp.synth({ aspectStabilization: stabilizeAspects });
+    } catch (error) {
+      if (error.message.includes('Cannot invoke Aspect')) {
+        return;
+      }
+      throw error;
+    }
+    try {
+      block(app);
+    } finally {
+      fs.rmSync(asm.directory, { recursive: true, force: true });
+    }
+  };
+}
+
+/**
+ * Implication operator, for readability
+ */
+function implies(a: boolean, b: boolean) {
+  return !a || b;
+}
+
+interface AspectVisitWithIndex extends AspectVisit {
+  index: number;
+}
+
+/**
+ * Check a property for every pair of visits in the log
+ *
+ * This is humongously inefficient at large scale, so we might need more clever
+ * algorithms to keep this tractable.
+ */
+function forEveryVisitPair(log: AspectVisitLog, block: (a: AspectVisitWithIndex, b: AspectVisitWithIndex) => void) {
+  for (let i = 0; i < log.length; i++) {
+    for (let j = 0; j < log.length; j++) {
+      if (i === j) { continue; }
+
+      block({ ...log[i], index: i }, { ...log[j], index: j });
+    }
+  }
+}
+
+function arraysEqual<T>(arr1: T[], arr2: T[]): boolean {
+  if (arr1.length !== arr2.length) {
+    return false;
+  }
+
+  return arr1.every((value, index) => value === arr2[index]);
+}
+
+function arraysEqualUnordered<T>(arr1: T[], arr2: T[]): boolean {
+  return arr1.length === arr2.length && new Set(arr1).size === new Set(arr2).size;
+}
+
+/**
+ * Given an AspectVisitLog, returns a map of Constructs with a list of all Aspects that
+ * visited the construct.
+ */
+function getVisitsMap(log: AspectVisitLog): Map<IConstruct, IAspect[]> {
+  const visitsMap = new Map<IConstruct, IAspect[]>();
+  for (let i = 0; i < log.length; i++) {
+    const visit = log[i];
+    if (!visitsMap.has(visit.construct)) {
+      visitsMap.set(visit.construct, []);
+    }
+    visitsMap.get(visit.construct)!.push(visit.aspect);
+  }
+  return visitsMap;
+}
+
+/**
+ * Returns a list of all nodes in the scope of the IConstruct.
+ */
+function getAllConstructsInScope(root: IConstruct): IConstruct[] {
+  const constructs: IConstruct[] = [root];
+
+  recurse(root);
+
+  function recurse(node: IConstruct) {
+    constructs.push(...node.node.children);
+    node.node.children.forEach(recurse);
+  }
```

Comment:
💅 The nicer form of recursion is:

```ts
const constructs = [];

  function recurse(node: IConstruct) {
    constructs.push(node);
    node.node.children.forEach(recurse);
  }
```

A recursion always has a "base case" and a "recurse case"

* base case: what do you do in case of a leaf
* recurse case: what do you do in case of a node.

In our case, if the root was a leaf, we would do nothing (having relied upon the fact that the setup would have placed the root into the result list already -- but that's then not part of the recursion).

It works, but it's a little harder to reason about imo.

You don't have to change it, because this whole function can be removed.